man/networkctl.xml

```diff
@@ -294,6 +294,35 @@ s - Service VLAN, m - Two-port MAC Relay (TPMR)
         which match the file are reconfigured.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term>
+          <command>create-netns</command>
+          <replaceable>NAME…</replaceable>
+        </term>
+        <listitem><para>Create a file referring to a Linux network namespace pseudo-file. Takes a name
+        or an absolute path. If a name is provided, then the file is created at
+        <filename>/var/run/netns/<replaceable>NAME</replaceable></filename>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term>
+          <command>remove-netns</command>
+          <replaceable>NAME…</replaceable>
+        </term>
+        <listitem><para>Remove a file referring to a Linux network namespace pseudo-file. Takes a name
+        or an absolute path. If a name is provided then it is assumed that the file is in
+        <filename>/var/run/netns</filename>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term>
+          <command>list-netns</command>
+          <optional><replaceable>PATH…</replaceable></optional>
+        </term>
+        <listitem><para>List files referring to Linux network namespace pseudo-files. If no path is
+        specified, then the files in <filename>/var/run/netns</filename> are shown.</para></listitem>
+      </varlistentry>
```

Comment:
These commands basically mirror the `ip netns list`, `ip netns add` and `ip netns delete` commands. Not sure if they bring additional value. One could argue `networkctl list` also is just a more fancy `ip l`, and we could show more information here, too.

I don't think there always needs to be a separate networkd instance for each namespace "managed" by networkd.

There's two usecases here:

1.  In addition to the default network namespace and main networkd, having a separate networkd running in a separate network namespace with interfaces in that namespace.
  Don't we have systemd-nspawn for this?

2.  Main networkd in the default network namespace sets up a tunnel interface (wireguard, openvpn), but moves the created interface into a separate network namespace (creating if it doesn't exist already), possibly configuring addresses/(default) route(s) on that interface as well.
  People that want to pipe traffic "through that tunnel" invoke applications (`ip netns exec $application`) or [whole graphical sessions](https://github.com/sddm/sddm/pull/798) in that namespace, but they don't really care about configuring more interfaces inside this namespace.

Reply:
> hmm, do we want these really? I think it would be much nicer if you allocate a new netns simply by starting another networkd instance. i.e.
> 
> "systemctl start [networkd@foobar.service](mailto:networkd@foobar.service)" starts a new netns called `foobar`

Let's remember that this really should fit well into regular workflow with "personalities". I would expect that:

```
ip netns exec alt-namespace-2 /bin/zsh
networkctl
```

returns the currently "logged in" network status, just like ip, iptables or tc tools are. Consider a scenario, when entire ssh server is being run within some netns (in general: one ssh server per netns) and one directly logs into the netns. It should be possible to manage this "internal" (separated) network, including stopping it - without interfering other namespaces.

Thinking about this makes me think, that maybe entire networkd@some-netns.service should be **started inside some-netns** (just like the ssh instance from example above), and the master (host) networkd.service should set up the namespaces itself and interfaces (like creating veth pair and moving one end into the specified netns). This seems to mean that .netdev with for NetNS=x would be picked by main systemd.service (in order to set up the device+netns assignment) **and** by networkd@x.service (in order to match the device), while .network for NetNS=x would belong to networkd@x.service only.

Then, the namespaced networkctl should display and manage only it's own parts, while the external one could give access to all the childrens. Also, some day, the systemctl could gain a feature to filter out namespace-bound services and disallow messing with master or siblings - this is the way of partitioning system and creating network-admin roles (known from Big Telco solutions; thinking about apache running within web1 netns and user/group X allowed to restart every service within web1 netns).

_As for the (simpified) use-case (italic, long): I need to set up 4 netnses: bgp1, bgp2, lan1, lan2 interconnected with veth pairs. The bgp* personalities would run separate BGP instances (bird) on assigned VLANs exchanging traffic with their respective lan* internal network BRASes. Each bgp* uses it's own uplinks by default and provides backup transmission for the other, so in general traffic can go through 3 namespaces (lan1-bgp1-bgp2), while in normal conditions the flows are separated and can be accounted per instance. In general setup there might be more than 2 "entity sets" and the "set" might contain more "boxes" (BGP, NAT/log, QoS/BRAS, IDS), while the "interface/VLAN" is stacked over some QinQ or LACP (802.1ad), VXLAN etc. Most of this can be acomplished by using "traditional" methods, like policy based routing, separate routing tables etc. but as the number of entities grow the configuration complicates exponentially as they start to heavily interfere. Using separate network namespaces, with their own routing tables, filtering rules (iptables), conntrack, DHCP servers etc. trades some performance for simplicity and manageability (like having different network-admins for different "sets"). Using full virtualization would result in much more performance penalty (and require additional layer of OpenVSwitch). Containers are functionally equivalent, but require massive system (fs tree) duplication and would became upgrade management hell really fast; note the network-admins should only care about the rules (BGP, iptables, ipsets, QoS), not the system ("firmaware") itself._